project/src/com/google/daggerquery/executor/models/Query.java

```diff
@@ -308,12 +302,50 @@ private void checkNodeForCorrectness(String node, BindingGraph bindingGraph) {
     List<String> closestNodes = new ArrayList<>();
 
     for (String node: allNodes) {
-      int distance = levenshteinDistanceCalculator.apply(originalNode, node);
+      int distance = calculateDistance(originalNode, node);
       if (distance <= MAX_NUMBER_OF_MISPLACED_LETTERS) {
         closestNodes.add(node);
       }
     }
 
     return closestNodes;
   }
+
+  /**
+   * Measures the difference between two strings.
+   *
+   * @return the number of changes required to get one string from another,
+   * where each change is a single character modification (substitution, deletion or insertion)
+   */
+  private int calculateDistance(String firstString, String secondString) {
+    int firstLength = firstString.length();
+    int secondLength = secondString.length();
+
+    int[][] distances = new int[firstLength + 1][secondLength + 1];
+
+    for (int index = 0; index <= firstLength; index++) {
+      distances[index][0] = index;
+    }
+
+    for (int index = 0; index <= secondLength; index++) {
+      distances[0][index] = index;
+    }
+
+    for (int firstIndex = 0; firstIndex < firstLength; firstIndex++) {
+      for (int secondIndex = 0; secondIndex < secondLength; secondIndex++) {
+        if (firstString.charAt(firstIndex) == secondString.charAt(secondIndex)) {
+          distances[firstIndex + 1][secondIndex + 1] = distances[firstIndex][secondIndex];
+        } else {
+          int replaceCost = distances[firstIndex][secondIndex] + 1;
+          int insertCost = distances[firstIndex][secondIndex + 1] + 1;
+          int deleteCost = distances[firstIndex + 1][secondIndex] + 1;
+
+          distances[firstIndex + 1][secondIndex + 1] =
+              Arrays.asList(replaceCost, insertCost, deleteCost).stream().min(Integer::compareTo).get();
```

Comment:
Can be simplified with [`Collections.min`](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#min-java.util.Collection-) :smiley:

Reply:
😄😄😄✔️